tests/integration/api/test_source.py

```diff
@@ -0,0 +1,223 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import uuid
+from dataclasses import dataclass
+from unittest import mock
+
+import pytest
+from rest_framework import status
+from rest_framework.test import APIClient
+
+from aap_eda.core import models
+from aap_eda.core.enums import Action, ResourceType
+from tests.integration.constants import api_url_v1
+
+
+@dataclass
+class InitData:
+    project: models.Project
+    activation: models.Activation
+    decision_environment: models.DecisionEnvironment
+    credential: models.Credential
+    rulebook: models.Rulebook
+    ruleset: models.Ruleset
+    rule: models.Rule
+    activation_instance: models.ActivationInstance
+
+
+@pytest.mark.django_db
+def test_list_sources(
+    client: APIClient, init_db, check_permission_mock: mock.Mock
+):
+    activation = init_db.activation
+
+    sources = models.Source.objects.bulk_create(
+        [
+            models.Source(
+                uuid=uuid.uuid4(),
+                name="test-source-1",
+                type="ansible.eda.range",
+                args={"limit": 5, "delay": 1},
+                activation=activation,
+                user=activation.user,
+                decision_environment=init_db.decision_environment,
+            ),
+            models.Source(
+                uuid=uuid.uuid4(),
+                name="test-source-2",
+                type="ansible.eda.range",
+                args={"limit": 6, "delay": 2},
+                activation=activation,
+                user=activation.user,
+                decision_environment=init_db.decision_environment,
+            ),
+        ]
+    )
+
+    response = client.get(f"{api_url_v1}/sources/")
+    assert response.status_code == status.HTTP_200_OK
+    assert len(response.data["results"]) == 2
+    response.data["results"][0]["uuid"] = sources[0].uuid
+    response.data["results"][0]["type"] = sources[0].type
+    response.data["results"][0]["name"] = sources[0].name
+    response.data["results"][0]["args"] = sources[0].args
+
+    check_permission_mock.assert_called_once_with(
+        mock.ANY, mock.ANY, ResourceType.SOURCE, Action.READ
+    )
+
+
+@pytest.mark.django_db
+def test_retrieve_source(
+    client: APIClient, init_db, check_permission_mock: mock.Mock
+):
+    activation = init_db.activation
+
+    source = models.Source.objects.create(
+        uuid=uuid.uuid4(),
+        name="test-source-1",
+        type="ansible.eda.range",
+        args={"limit": 5, "delay": 1},
+        activation=activation,
```

Comment:
@hsong-rh I am thinking this will happen later the user won't be passing in an activation we will create it behind the scenes.

Reply:
We can leave it as Null and have the post create task that creates and starts the activation fill it in.

Reply:
Agree